spec/models/odk_config_generator_spec.rb

```diff
@@ -0,0 +1,37 @@
+require 'spec_helper'
+
+describe ODKConfigGenerator do
+  describe "#generate_string" do
+    it "returns valid json string" do
+      test_username = "test_user"
+      test_pwd = "test_pwd"
+      site_url = "https://secure.cceom.org/m/kenya2017"
+      expected = '{"general":{"password":"test_pwd","username":"test_user","server_url":"https://secure.cceom.org/m/kenya2017"},"admin":{}}'
+      actual = ODKConfigGenerator.new().generate_string(test_username, test_pwd, site_url)
+
+      expect(actual).to eq expected
+    end
+  end
+
+  describe "#generate_odk_config" do
+    it "calls generate_string" do
+      subject = ODKConfigGenerator.new()
+      expect(subject).to receive(:generate_string).and_return("string")
+
+      subject.generate_odk_config("test", "test", "test")
+    end
+
+    it "deflates a string" do
+      expect(Zlib::Deflate).to receive(:deflate).and_return(Zlib::Deflate.deflate("string"))
+
+      ODKConfigGenerator.new().generate_odk_config("test", "test", "test")
+    end
+
+    it "encodes the string" do
+      return_value = Base64.strict_encode64(Zlib::Deflate.deflate("string"))
+      expect(Base64).to receive(:strict_encode64).and_return(return_value)
+
+      ODKConfigGenerator.new().generate_odk_config("test", "test", "test")
+    end
+  end
```

Comment:
These specs are nice and thorough but really I don't think you need this much. 

All you really need to test is that you're generating the correct base64 string for a given url, username, and password. We can assume the QRCode library is correct.

What I would do is compute outside of Ruby a correct base64 string and put it in here. (Since you already have this all working, you can cheat a bit and dump out a valid one from the Rails console and use that as your ground truth since you know it's correct.) Then add an accessor method to your class that returns the computed base64 string you're generating. Then compare the two in the spec.

I'm not sure that we gain anything by testing that deflate gets called or that encode gets called or etc. I think that's a bit too fine grained and if you always write tests at that level it will really slow you down. If you get the test I described above passing, you know all of those things must be happening.

Also in the above test you're not actually testing e.g. if zip and encode are called in the right order. I think if you flipped them in your code, the tests would still pass.

I know this is a lot. Picking out the right things to test is one of the hardest things to do in TDD and so I want to be sure we're on the same page on this. Happy to discuss it tomorrow.